### PR TITLE
docs(proxy): add ssl guidelines for communication between Agent and Datadog when using a proxy

### PR DESCRIPTION
**Pull request by Kaderinho**
### What does this PR do?

- Add guidelines on raw/encrypted communication between the Agent,  proxy and Datadog intake.
    - Add Datadog certificate verification on the proxy host.
    - Add docs on when and how to encrypt communication between the Agent and the proxy
- Fix some errors :
    - telemetry HAProxy frontend was missing after a merge
    - AppSec port was different on HAProxy and Nginx + missing in one `datadog.yaml` config
    - update the part about `skip_ssl_validation` (which was previously useless because HTTP was used between the Agent and the proxy)

### Motivation

Give better guidelines to customers

https://docs-staging.datadoghq.com/nicolas.guerguadj/aml-100-add-https-guideline-for-proxy-setup/agent/proxy/?tab=agentv6v7

### Additional Notes

- Did not test `network_devices` but as it is using the same component as `database_monitoring`, the behavior is the same
- There is still no guidelines on how to create certificates, will be addressed in a future PR

### To check

There was some modifications on this file before this PR that introduced some errors like missing frontend or inconsistencies with ports. I realized it and fixed them after all I added, so I might have missed some stuff : 

It would be nice if someone check after me that : 

- [ ]  `HTTP` configuration (the one by default before this PR) should only have : 
    - [ ] `ssl verify <PATH_TO_CERTIFICATES>` instead of `ssl verify none` for HAProxy
    - [ ] `proxy_ssl_verify on` added to all server for NGINX
- [ ]  `HTTPS` configuration should contains the `HTTP` config information and : 
    - [ ] `ssl crt <PATH_TO_PROXY_CERTIFICATE>` at the end on each `bind` line for HAProxy
    - [ ] `ssl` at the end on each `listen` line for NGINX
- [ ] There is no `HAProxy` reference in `NGINX` part and vice versa 

---

### Reviewer checklist
- [ ] Review the changed files.
- [ ] Review the URLs listed in the [Preview](#preview) section.
- [ ] Check images for PII
- [ ] Review any mentions of "Contact Datadog support" for internal support documentation.

### CODE REVIEW

**Comment by maycmlee** [comment]
Created DOCS-3856 for editorial review.

**Review by djmitche — COMMENTED** [review]
I don't know enough about the context to give this a 👍🏽 but at a high level it makes sense to me.

File: content/en/agent/proxy.md
Comment (Kaderinho) [inline content/en/agent/proxy.md:540]:
```suggestion
# sending Instrumentation Telemetry data (e.g. the value of "apm_config.telemetry.dd_url")
```

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:160]: We should add a note here and in https version for users to replace all `<VARIABLE_NAME>` sections to prevent them from blindly copying them. This should probably be reflected in NGINX doc wording too.

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:468]: Not sure about the name here since it should also imply that it has a key with it. Maybe this?
```suggestion
    bind *:3835 ssl crt <PATH_TO_PROXY_CERTIFICATE_PEM>
```

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:664]: The download of backend certificates should probably be mentioned before the config if we keep the `PATH_TO_CERTIFICATES` in the first example.

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:673]:
```suggestion
**Note**: You can use `ssl verify none` instead of `ssl verify <PATH_TO_CERTIFICATES>` if you are unable to get the certificates on the proxy host, but be aware that HAProxy will not be able to verify Datadog's intake certificate in that case.
```

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:688]: We should add that the scheme needs to match their HAProxy settings

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:737]:
```suggestion
When using encryption between the Agent and HAProxy, if the Agent does not have access to the proxy certificate, is unable to validate it, or the validation is not needed, you can edit the `datadog.yaml` Agent configuration file and set `skip_ssl_validation` to `true`.
```

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:738]:
```suggestion
With this option set to `true`, the Agent will skip the certificate validation step so it will not verify the identity of the proxy but the communication will still be encrypted with SSL/TLS.
```

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:812]:
```suggestion
**Note**: It is recommended to use the `HTTPS` configuration file if the Agent and NGINX are not part of the same local isolated network.
```

File: content/en/agent/proxy.md
Comment (sgnn7) [reply content/en/agent/proxy.md:151]: This should probably be reflected in NGINX doc wording too

**Review by sgnn7 — COMMENTED** [review]
@Kaderinho Good work on improving these docs! I left some comments and feel free to ping me on them as-needed since most of them are subjective to a degree.

File: content/en/agent/proxy.md
Comment (Kaderinho) [reply content/en/agent/proxy.md:152]: > Isn't that just a regular SSL/TLS certificate? Do we mean an x509 certificate?

Yeah, we mean an x509 certificate here but if I correctly understand, there is a field named `Common Name` (or CN). which is deprecated by the [RFC 2818](https://datatracker.ietf.org/doc/html/rfc2818#section-3.1) in 2000 and I believe you can still create certificate using `CN` instead of `SAN`.

```
If a subjectAltName extension of type dNSName is present, that MUST be used as the identity. 
...
Although the use of the Common Name is existing practice, it is deprecated 
and Certification Authorities are encouraged to use the dNSName instead
```

 I added this precision because the Agent throws an error when trying to verify a certificate with CN and asks for SAN instead. Maybe should I say `you will need to create a x509 certificate with Subject Alternative Name (SAN) extension` ?

---

> I think this can be reworded just to say that "you will need to create a Subject Alternative Name (SAN) certificate and its private key for the HAProxy host" ? 

The problem here is that NGINX has 2 attributes : 

-  `ssl_certificate`  for the certificate (usually `.crt`)
-  `ssl_certificate_key` for the key (usually `.key`)

But HAProxy has only one attribute `ssl crt <cert.pem>` where `<cert.pem>` is a file containing both the certificate and the private key (can be created with `cat certificate.crt private_key.key > cert.pem`). I wanted to emphasise the fact that the certificate and private key must be in the same file but I agree that it is unclear so feel free to give any suggestions on how to rephrase it 🙏

File: content/en/agent/proxy.md
Comment (Kaderinho) [reply content/en/agent/proxy.md:312]: Do you think that keeping `ssl verify none` in the configuration and add explanation on how to verify Datadog certificate would be better ?

> This may be something to cover with a separate config type as well.

Do you mean with new tabs ?

File: content/en/agent/proxy.md
Comment (sgnn7) [reply content/en/agent/proxy.md:152]: > Maybe should I say `you will need to create a x509 certificate with Subject Alternative Name (SAN) extension`

You can't have a SAN field without it being an x509 certificate I believe, this works great - makes it clear about what they need to do. 

> feel free to give any suggestions on how to rephrase it

Maybe we can call it a certificate bundle? WDYT?
```
This certificate bundle (*.pem) should contain both the public certificate and private key. See this [HAProxy blog post][3] for more information.
```

File: content/en/agent/proxy.md
Comment (sgnn7) [reply content/en/agent/proxy.md:312]: > Do you think that keeping ssl verify none in the configuration and add explanation on how to verify Datadog certificate would be better ?

To me that would be the least friction for adopters but I can't comment on it confidently since I don't know for sure what the specific intents of the effort are.

> Do you mean with new tabs ?

Just mean that we can have a config with `ssl verify none` and then one with either an explanation (like you mentioned) or a full new config that has verification of the backend.

Comment (Kaderinho) [reply content/en/agent/proxy.md:312]: Honestly I don't really know what to think about this part. I feel like we should prioritize 'security' and provide 'correct' guidelines (actual ones have security issues) even if it increases the setup / maintenance complexity.

But I completely understand the fact that we must keep the configuration simple. Imo the current trade-off is : 

## With the certificate verification 
- Pros :
     - More accurate guidelines / more secure
     - More awareness from customers about certificates verification (as they need to setup it)
- Cons :
     - More complexity on the setup / maintenance 

## Without verification 
- Pros : 
    -  Easy to setup
- Cons : 
    - Not optimal guidelines 
    -  Customers might not be aware of the provided configuration problems

Did I miss something ? 

My opinion is that I would let the configuration as it and add more details about how to do deal with `<PATH_TO_CERIFICATES>`  (they can still replace it by `none` as an alternative). But you have a better vision than me about what is better for customers and how you want the proxy guidelines to evolve in the future so feel free to decide.

File: content/en/agent/proxy.md
Comment (Kaderinho) [inline content/en/agent/proxy.md:1037]:
```suggestion
**Note**: You can remove `proxy_ssl_verify on` if you are unable to get the certificates on the proxy host, but be aware that NGINX will not be able to verify Datadog's intake certificate in that case.
```

File: content/en/agent/proxy.md
Comment (Kaderinho) [inline content/en/agent/proxy.md:1096]:
```suggestion
When using encryption between the Agent and NGINX, if the Agent does not have access to the proxy certificate, is unable to validate it, or the validation is not needed, you can edit the `datadog.yaml` Agent configuration file and set `skip_ssl_validation` to `true`.
```

File: content/en/agent/proxy.md
Comment (Kaderinho) [inline content/en/agent/proxy.md:1097]:
```suggestion
With this option set to `true`, the Agent will skip the certificate validation step so it will not verify the identity of the proxy but the communication will still be encrypted with SSL/TLS.
```

File: content/en/agent/proxy.md
Comment (Kaderinho) [reply content/en/agent/proxy.md]: Done here : cca6a1513774cf4d69a0b2d28a621cbba772027e

File: content/en/agent/proxy.md
Comment (Kaderinho) [reply content/en/agent/proxy.md]: I rephrased it here : 41ea6a4696acb87e90d0bef3aa1899e4a3ab4669 WDYT ?

**Comment by sgnn7** [comment]
Hey @Kaderinho, I briefly mentioned the new `ssl verify <path>` guidance  to @olivielpeau today as it  impacts the FIPS proxy deployments as well but we didn't have enough time to resolve the path forward. We have a meeting tomorrow (Wed, Aug 3) on figuring out how we want to handle the `ssl verify ...` and I'll do a follow-up review after that. Sorry about the delay!

File: content/en/agent/proxy.md
Comment (sgnn7) [inline content/en/agent/proxy.md:1033]:
```suggestion
**Note**: Install the common root CA certificates with one of the following commands:
```

File: content/en/agent/proxy.md
Comment (sgnn7) [reply content/en/agent/proxy.md:666]:
```suggestion
**Note**: Install the common root CA certificates with one of the following commands:
```

**Review by sgnn7 — APPROVED** [review]
@Kaderinho 3 tiny nits left but LGTM!

PS: Olivier and I agreed that this improves our docs regardless of ease-of-use concerns mentioned earlier so nothing is blocking the merge of this PR.

File: content/en/agent/proxy.md
Comment (hestonhoffman) [inline content/en/agent/proxy.md:923]:
```suggestion
This configuration adds SSL/TLS encryption on communication between the Agent and NGINX. Replace `<PATH_TO_PROXY_CERTIFICATE>` with the path to the proxy public certificate and `<PATH_TO_PROXY_CERTIFICATE_KEY>` with the path to the private key.
```

File: content/en/agent/proxy.md
Comment (hestonhoffman) [reply content/en/agent/proxy.md:1033]: Move this section above the config

File: content/en/agent/proxy.md
Comment (hestonhoffman) [inline content/en/agent/proxy.md:1051]:
```suggestion
Replace `<SCHEME>` with `https` if you previously chose the HAProxy HTTPS configuration, or with `http` if you did not choose HTTPS.
```

File: content/en/agent/proxy.md
Comment (hestonhoffman) [inline content/en/agent/proxy.md:1102]:
```suggestion
With this option set to `true`, the Agent skips the certificate validation step and does not verify the identity of the proxy, but the communication is still encrypted with SSL/TLS.
```

File: content/en/agent/proxy.md
Comment (hestonhoffman) [inline content/en/agent/proxy.md:1159]:
```suggestion
[4]: https://www.nginx.com
[5]: /agent/logs/proxy
```

File: content/en/agent/proxy.md
Comment (hestonhoffman) [inline content/en/agent/proxy.md:1108]:
```suggestion
When sending logs over TCP, see [TCP Proxy for Logs][5].
```

**Comment by Kaderinho** [comment]
Hi @hestonhoffman,

Sorry for the delay 🙇 I have applied the suggestions (04e253865ed771231eb07331eed5003768a9764d) and fix an error in the HAProxy config (3919cfe80e4b9a481d8883649ecb1fa5663c1230).

I'll do a final check tomorrow morning to verify one last time that both configuration are correct (just in case) but if everything is good for you, please feel free to approve and I'll merge it tomorrow 🙏

**Review by hestonhoffman — APPROVED** [review]
A couple small changes, but this is ready to go once those are merged. I think you need a code owner to merge. I'll check back in tomorrow

**Comment by Kaderinho** [comment]
I applied suggestions and correct one file (missing `;`), I checked HTTP and HTTPS configuration for both HAProxy and NGINX and that LGTM

Thanks for the review !

**Review by hestonhoffman — APPROVED** [review]

**Pull request merged**
